src/annotator/integrations/pdf.js

```diff
@@ -309,14 +334,81 @@ export class PDFIntegration {
   }
 
   /**
+   * Scroll to the location of an anchor in the PDF.
+   *
+   * If the anchor refers to a location that is an un-rendered page far from
+   * the viewport, then scrolling happens in three phases. First the document
+   * scrolls to the approximate location indicated by the placeholder anchor,
+   * then `scrollToAnchor` waits until the page's text layer is rendered and
+   * the annotation is re-anchored in the fully rendered page. Then it scrolls
+   * again to the final location.
+   *
    * @param {Anchor} anchor
    */
-  scrollToAnchor(anchor) {
-    // TODO - Account for pages changing state between un-rendered and rendered
-    // while scrolling.
-    const highlights = /** @type {Element[]} */ (anchor.highlights);
-    return new Promise(resolve => {
-      scrollIntoView(highlights[0], resolve);
-    });
+  async scrollToAnchor(anchor) {
+    const annotation = anchor.annotation;
+    const inPlaceholder = anchorIsInPlaceholder(anchor);
+    const offset = this._anchorOffset(anchor);
+    if (offset === null) {
+      return;
+    }
+
+    // nb. We only compute the scroll offset once at the start of scrolling.
+    // This is important as the highlight may be removed from the document during
+    // the scroll due to a page transitioning from rendered <-> un-rendered.
+    await scrollElement(this.contentContainer(), offset);
```

Comment:
https://github.com/hypothesis/client/issues/3269 notes a sporadic problem where the PDF would sometimes jump right to the top after clicking on an annotation card:

> Additionally if you repeat steps 3 and 4 a few times you may also encounter times when the viewer starts scrolling and then reverts to scrolling back up to the top.

The cause of this, prior to this PR, is when the highlight element got removed from the DOM mid-scroll, causing `scrollIntoView` to compute its offset as 0.

Reply:
Golly. Good catch. I could see this causing a hard-to-track-down bug, yeah.